Changes to allow predicting/relaxing with NVIDIA L4 GPUs

cjpooley opened the pull request:
# Description
- Upgraded CUDA libraries (see https://en.wikipedia.org/wiki/CUDA#GPUs_supported), Python (3.8->3.9) and JAX libs
- Modified pipeline to create custom training job components to support specifying machine type/GPU type (see https://cloud.google.com/vertex-ai/docs/pipelines/request-gcp-machine-resources), which is not supported with the default component SELECTOR_CONSTRAINT (see https://cloud.google.com/vertex-ai/docs/pipelines/machine-types)
- Modified config.py to replace old component params with new
- Modified predict and relax component definitions to incorporate XLA_PYTHON_CLIENT_MEM_FRACTION and TF_FORCE_UNIFIED_MEMORY since specifying on component instances (in the pipeline) as done previously does not carry through to the underlying component container.
- Fix bug on `predict_relax` component, the parameter run_relax expects a boolean.

# How has this been tested?

Tested on my internal GCP project, across all kinds of sequences and GPU types T4 and L4. Leaving A100 for another PR, if required.

# Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have successfully run the E2E tests, and have included the links to the pipeline runs below
- [x] New and existing unit tests/use cases pass locally with my changes
- [x] I have updated any relevant documentation to reflect my changes
- [x] I have assigned a reviewer and messaged them

Commit:
Replicating changes

Commit:
Updated env vars

Commit:
Updates to env vars and relax config

JasperW01 commented:
Great pull request and also very timely. 

Is it possible to make accelerator_count for Predict & Relax also configurable while the default is "1"? It's quite common to configure multiple GPUs (such as A100) to deal with complex proteins with big residue number. 

Thx and cheers.

chmstimoteo commented on 1-alphafold-quick-start.ipynb:
```diff
@@ -181,16 +181,12 @@
     "# os.environ['DATA_PIPELINE_MACHINE_TYPE'] = 'c2-standard-16'\n",
     "\n",
     "# # Instance (VM) configuration to run model prediction\n",
-    "# os.environ['MEMORY_LIMIT'] = '85'    # Amount of host memory (RAM)\n",
```
I don't think there is a need to delete these... leave them commented

chmstimoteo commented on src/config.py:
```diff
@@ -49,16 +49,11 @@
 HMMSEARCH_MACHINE_TYPE = os.getenv('HMMSEARCH_MACHINE_TYPE', 'c2-standard-16')
 HHBLITS_MACHINE_TYPE = os.getenv('HMMSEARCH_MACHINE_TYPE', 'c2-standard-16')
 
-MEMORY_LIMIT = os.getenv('MEMORY_LIMIT', '32')
```
Could we keep the same set of env variables for the 2 different tasks (predict and relax)?

chmstimoteo commented on src/pipelines/alphafold_inference_pipeline.py:
```diff
@@ -94,38 +109,26 @@ def alphafold_inference_pipeline(
-    model_predict.set_gpu_limit(config.GPU_LIMIT)
-    model_predict.add_node_selector_constraint(
-        config.GKE_ACCELERATOR_KEY, config.GPU_TYPE)
-    model_predict.set_env_variable(
```
Why deleting these? It makes it not retro-compatible with the previous pipeline version, in the case there is no need to use the new environment variables and parameters.

chmstimoteo replied:
As per another the author reply on another change, `.set_env_variable()` doesn't work. Resolving this item.

chmstimoteo commented on src/pipelines/alphafold_inference_pipeline.py:
```diff
@@ -94,38 +109,26 @@ def alphafold_inference_pipeline(
-      relax_protein.set_gpu_limit(config.RELAX_GPU_LIMIT)
-      relax_protein.add_node_selector_constraint(
-        config.GKE_ACCELERATOR_KEY, config.RELAX_GPU_TYPE)
-      relax_protein.set_env_variable(
```
I still believe using the .set_env_variable() being a good practice, unless during test those variables are not hydrated to the google-cloud-pipeline-components (v1.0.44) custom job op machine and container.

cjpooley replied:
Thanks for all the feedback @chmstimoteo. Will discuss live with you next week as discussed. Understood on the backward compatibility concern. As far as env variables: Because there's a [custom Vertex training job component](https://cloud.google.com/vertex-ai/docs/pipelines/request-gcp-machine-resources) wrapping the custom job op component (required to allow specifying machine type/GPU type), env vars specified using .set_env_variable() don't make it to the underlying component based on my testing.

JasperW01 replied:
Yeah, I was also puzzled by the same problem. Last month when I was trying L4 with create_custom_training_job_from_component. The env setting by .set_env_variable() seems not working any more.  Looks like the approach in this Pull Request to set env variables is the only way.

chmstimoteo reviewed (changes requested):
Comments on files.

chmstimoteo commented:
> Great pull request and also very timely.
> 
> Is it possible to make accelerator_count for Predict & Relax also configurable while the default is "1"? It's quite common to configure multiple GPUs (such as A100) to deal with complex proteins with big residue number.
> 
> Thx and cheers.

Agreed with @JasperW01 .

Commit:
- Added accelerator count env variable
- Updated optimized monomer and sequential pipelines to use new variables
- Resolved bug in sequential pipeline with type mismatch

cjpooley commented:
> Great pull request and also very timely.
> 
> Is it possible to make accelerator_count for Predict & Relax also configurable while the default is "1"? It's quite common to configure multiple GPUs (such as A100) to deal with complex proteins with big residue number.
> 
> Thx and cheers.

Thanks @JasperW01, and great suggestion. Adding this in.

JasperW01 commented on Dockerfile:
```diff
@@ -60,10 +60,10 @@ ENV PATH="/opt/conda/bin:$PATH"
       pip \
-      python=3.8 \
+      python=3.9 \
+      && conda install -y -c nvidia/label/cuda-11.8.0 cuda-libraries-dev cuda-nvcc cuda-nvtx cuda-cupti \
```
Hi @cjpooley , just curious, in this line, why we don't have cuda-toolkit which would be equivalent to previous version ? Instead it has a series of multiple cuda components?

cjpooley replied:
There was a change made to the cudatoolkit so that at least by version 11.8 it no longer included required components, including nvcc. This caused errors in my initial testing, so I followed the guidance [here](https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html#installing-cuda-using-conda:~:text=If%20you%20had%20previously%20installed%20CUDA%20using%20the%20cudatoolkit%20package%20and%20want%20to%20maintain%20a%20similar%20install%20footprint%2C%20you%20can%20limit%20your%20installation%20to%20the%20following%20packages%3A).

Thanks for the suggestion on referencing the CUDA arg - will do.

JasperW01 replied:
Not sure if you have A100 quota handy to replicate it.

JasperW01 commented:
Hi @cjpooley , 

I tested the Pull request with both L4 & A100 for the same protein of https://github.com/GoogleCloudPlatform/vertex-ai-alphafold-inference-pipeline/blob/main/sequences/1S78.fasta. Although it works with L4 but it failed with A100. 

With A100, the predict step is successful and the relax predict step started with the problem of "Error loading CUDA module: CUDA_ERROR_UNSUPPORTED_PTX_VERSION (222)", then it gave up and failed after 100 times of retry. 

screenshot is below

<img width="1225" alt="image" src="https://github.com/GoogleCloudPlatform/vertex-ai-alphafold-inference-pipeline/assets/14971090/393251b1-f922-4a3a-8fe1-ac72ac4ed55d">

<img width="1226" alt="image" src="https://github.com/GoogleCloudPlatform/vertex-ai-alphafold-inference-pipeline/assets/14971090/ac66f679-c32d-4246-90ad-ce14ebce2153">

cjpooley commented:
> Hi @cjpooley ,
> 
> I tested the Pull request with both L4 & A100 for the same protein of https://github.com/GoogleCloudPlatform/vertex-ai-alphafold-inference-pipeline/blob/main/sequences/1S78.fasta. With L4, it works E2E with a couple of warning messages in the Predict step which I am still digging.
> 
> With A100, the Predict step works well with no warning but the Relax step is hit with "Error loading CUDA module: CUDA_ERROR_UNSUPPORTED_PTX_VERSION (222)", then after 100 times of retries it gave up and failed.
> 
> screenshot is below
> 
> <img alt="image" width="1225" src="https://user-images.githubusercontent.com/14971090/268534069-393251b1-f922-4a3a-8fe1-ac72ac4ed55d.png"> <img alt="image" width="1226" src="https://user-images.githubusercontent.com/14971090/268534095-ac66f679-c32d-4246-90ad-ce14ebce2153.png">
> not sure if you have A100 quota handy to replicate it or not.

Thank you for raising this (and testing with the A100.) I will look into this as soon as I can.

chmstimoteo commented on 1-alphafold-quick-start.ipynb:
```diff
@@ -181,16 +181,12 @@
-    "# os.environ['RELAX_MEMORY_LIMIT'] = '85'    # Amount of host memory (RAM)\n",
-    "# os.environ['RELAX_CPU_LIMIT'] = '12'       # Number of CPUs\n",
-    "# os.environ['RELAX_GPU_LIMIT'] = '1'        # Number of GPUs\n",
-    "# os.environ['RELAX_GPU_TYPE'] = 'nvidia-tesla-a100'  # GPU type"
```
we get rid

chmstimoteo reviewed (approved):
LGTM!

JasperW01 commented:
HI @cjpooley , 

Just add an clarification to my comment above which has screenshot - all those contents are correct except I missed one point - to make L4 successfully predict 1S78, I had to consolidate the recent changes in the [Dockerfile](https://github.com/google-deepmind/alphafold/blob/main/docker/Dockerfile) from the main branch of Deepmind AlphaFold github repo into your Dockerfile. Otherwise, predicting 1S78 with L4 fails at the Predict step. 

Happy to jump on a call to show you at your convenient time.

JasperW01 commented:
This is the Dockerfile I used which can predict 1S78 with L4 successfully but fails with A100 at the Relax step. And as you can see, it's basically on top of your Dockerfile but merged with recent changes from the Deepmina github repo's Dockerfile. 

# Copyright 2022 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

ARG CUDA=11.8.0
FROM nvidia/cuda:${CUDA}-cudnn8-runtime-ubuntu18.04
# FROM directive resets ARGS, so we specify again (the value is retained if
# previously set).
ARG CUDA
ARG ALPHAFOLD_VERSION=main

# Use bash to support string substitution.
SHELL ["/bin/bash", "-o", "pipefail", "-c"]

RUN apt-get update \ 
    && DEBIAN_FRONTEND=noninteractive apt-get install --no-install-recommends -y \
        build-essential \
        cmake \
        cuda-command-line-tools-$(cut -f1,2 -d- <<< ${CUDA//./-}) \
        git \
        hmmer \
        kalign \
        tzdata \
        wget \
    && rm -rf /var/lib/apt/lists/* \
    && apt-get autoremove -y \
    && apt-get clean

# Compile HHsuite from source.
RUN git clone --branch v3.3.0 https://github.com/soedinglab/hh-suite.git /tmp/hh-suite \
    && mkdir /tmp/hh-suite/build \
    && pushd /tmp/hh-suite/build \
    && cmake -DCMAKE_INSTALL_PREFIX=/opt/hhsuite .. \
    && make -j 4 && make install \
    && ln -s /opt/hhsuite/bin/* /usr/bin \
    && popd \
    && rm -rf /tmp/hh-suite

# Install Miniconda package manager.
RUN wget -q -P /tmp \
  https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh \
    && bash /tmp/Miniconda3-latest-Linux-x86_64.sh -b -p /opt/conda \
    && rm /tmp/Miniconda3-latest-Linux-x86_64.sh

# Install conda packages.
ENV PATH="/opt/conda/bin:$PATH"
##  Bug-Fix: Running the conda version 4.13.0, we face an incompatibility issue with python 3.11. 
##  Issue and fix appled as suggested in this github issue: https://github.com/deepmind/alphafold/issues/798
#RUN conda install -qy conda==4.13.0 \
RUN conda install -qy conda==23.5.2 \
    && conda install -y -c conda-forge \
      openmm=7.7.0 \
      pdbfixer \
      pip \
      python=3.10 \
      && conda install -y -c "nvidia/label/cuda-${CUDA}" cuda-toolkit \
      && conda clean --all --force-pkgs-dirs --yes

RUN git clone --branch $ALPHAFOLD_VERSION https://github.com/deepmind/alphafold.git /app/alphafold
RUN wget -q -P /app/alphafold/alphafold/common/ \
  https://git.scicore.unibas.ch/schwede/openstructure/-/raw/7102c63615b64735c4941278d92b554ec94415f8/modules/mol/alg/src/stereo_chemical_props.txt

# Install pip packages.
RUN pip3 install --upgrade pip --no-cache-dir \
    && pip3 install -r /app/alphafold/requirements.txt --no-cache-dir \
    && pip3 install --upgrade --no-cache-dir \
      jax==0.4.13 \
      jaxlib==0.4.13+cuda11.cudnn86 \
      -f https://storage.googleapis.com/jax-releases/jax_cuda_releases.html

# Add SETUID bit to the ldconfig binary so that non-root users can run it.
RUN chmod u+s /sbin/ldconfig.real

WORKDIR /modules
ADD src/components/alphafold_utils.py .

ENV PYTHONPATH=/app/alphafold:/modules
RUN ldconfig

Commit:
Parameterizing CUDA version in Dockerfile

chmstimoteo commented:
> Hi @cjpooley ,
> 
> I tested the Pull request with both L4 & A100 for the same protein of https://github.com/GoogleCloudPlatform/vertex-ai-alphafold-inference-pipeline/blob/main/sequences/1S78.fasta. With L4, it works E2E with a couple of warning messages in the Predict step which I am still digging.
> 
> With A100, the Predict step works well with no warning but the Relax step is hit with "Error loading CUDA module: CUDA_ERROR_UNSUPPORTED_PTX_VERSION (222)", then after 100 times of retries it gave up and failed.
> 
> screenshot is below
> 
> <img alt="image" width="1225" src="https://user-images.githubusercontent.com/14971090/268534069-393251b1-f922-4a3a-8fe1-ac72ac4ed55d.png"> <img alt="image" width="1226" src="https://user-images.githubusercontent.com/14971090/268534095-ac66f679-c32d-4246-90ad-ce14ebce2153.png">
> not sure if you have A100 quota handy to replicate it or not.

@JasperW01 we're going to adress compatibility with A100 on another PR.
